src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java

```diff
@@ -185,34 +189,31 @@ public SpawnResult exec(Spawn spawn, SpawnExecutionContext context)
         return execLocally(spawn, context, inputMap, uploadLocalResults, remoteCache, actionKey);
       }
 
+      ExecuteRequest request = ExecuteRequest.newBuilder()
+          .setInstanceName(options.remoteInstanceName)
+          .setAction(action)
+          .setSkipCacheLookup(!acceptCachedResult)
+          .build();
       try {
-        // Upload the command and all the inputs into the remote cache.
-        remoteCache.ensureInputsPresent(repository, execRoot, inputRoot, command);
-      } catch (IOException e) {
-        return execLocallyOrFail(spawn, context, inputMap, actionKey, uploadLocalResults, e);
-      }
+        try {
+          return retrier.execute(() -> {
+            // Upload the command and all the inputs into the remote cache.
+            remoteCache.ensureInputsPresent(repository, execRoot, inputRoot, command);
 
-      final ActionResult result;
-      boolean remoteCacheHit = false;
-      try {
-        ExecuteRequest.Builder request =
-            ExecuteRequest.newBuilder()
-                .setInstanceName(options.remoteInstanceName)
-                .setAction(action)
-                .setSkipCacheLookup(!acceptCachedResult);
-        ExecuteResponse reply = remoteExecutor.executeRemotely(request.build());
-        maybeDownloadServerLogs(reply, actionKey);
-        result = reply.getResult();
-        remoteCacheHit = reply.getCachedResult();
-      } catch (IOException e) {
-        return execLocallyOrFail(spawn, context, inputMap, actionKey, uploadLocalResults, e);
-      }
+            ExecuteResponse reply = remoteExecutor.executeRemotely(request);
+            maybeDownloadServerLogs(reply, actionKey);
 
-      try {
-        return downloadRemoteResults(result, context.getFileOutErr())
-            .setRunnerName(remoteCacheHit ? "remote cache hit" : getName())
-            .setCacheHit(remoteCacheHit)
-            .build();
+            return downloadRemoteResults(reply.getResult(), context.getFileOutErr())
```

Comment:
This method already does retries. I think nesting retries is probably best avoided if for no other reason but to keep the code simple. What do you think about moving this method outside the retrier?

Reply:
executeRemotely also does retries (another two levels of them, actually), but the point of this is to retry this particular sequence based on transient failures that can only be guaranteed by restarting the sequence: FAILED_PRECONDITION and CacheNotFoundException mean that the inputs or successfully executed action's outputs respectively have been lost and the action must be reexecuted. No other level retries on these exception types, and these are both non-retriable exceptions for the request-level retriers at lower levels.

This is all about surviving CAS/AC/Ex transiency and all of the possible state transitions for it as an unreliable service, and since any build *can* be in one of those states, they *will* be. I'll see to adding some colorful tests to indicate the variety of conditions I encountered on the way to this.